src/FindIntrinsics.cpp

```diff
@@ -0,0 +1,714 @@
+            auto is_x_same_uint = op->type.is_uint() && is_uint(x, bits);
+            auto is_x_same_int_or_uint = is_x_same_int || is_x_same_uint;
+            // clang-format off
+            if (rewrite(max(min(widening_add(x, y), upper), lower), saturating_add(x, y), is_x_same_int_or_uint) ||
+                rewrite(max(min(widening_sub(x, y), upper), lower), saturating_sub(x, y), is_x_same_int_or_uint) ||
+                rewrite(min(widening_add(x, y), upper), saturating_add(x, y), op->type.is_uint() && is_x_same_uint) ||
+                rewrite(max(widening_sub(x, y), lower), saturating_sub(x, y), op->type.is_uint() && is_x_same_uint) ||
+
+                rewrite(shift_right(widening_add(x, y), 1), halving_add(x, y), is_x_same_int_or_uint) ||
+                rewrite(shift_right(widening_sub(x, y), 1), halving_sub(x, y), is_x_same_int_or_uint) ||
+
+                rewrite(halving_add(widening_add(x, y), 1), rounding_halving_add(x, y), is_x_same_int_or_uint) ||
+                rewrite(halving_add(widening_add(x, 1), y), rounding_halving_add(x, y), is_x_same_int_or_uint) ||
+                rewrite(halving_add(widening_sub(x, y), 1), rounding_halving_sub(x, y), is_x_same_int_or_uint) ||
+                rewrite(rounding_shift_right(widening_add(x, y), 1), rounding_halving_add(x, y), is_x_same_int_or_uint) ||
+                rewrite(rounding_shift_right(widening_sub(x, y), 1), rounding_halving_sub(x, y), is_x_same_int_or_uint) ||
+
+                // We can ignore the sign of the widening subtract for halving subtracts.
+                rewrite(shift_right(cast(op_type_wide, widening_sub(x, y)), 1), halving_sub(x, y), is_x_same_int_or_uint) ||
+                rewrite(rounding_shift_right(cast(op_type_wide, widening_sub(x, y)), 1), rounding_halving_sub(x, y), is_x_same_int_or_uint) ||
+
+                false) {
+                return mutate(rewrite.result);
+            }
+            // clang-format on
+
+            // When the argument is a widened rounding shift, we might not need the widening.
+            // When there is saturation, we can only avoid the widening if we know the shift is
+            // a right shift. Without saturation, we can ignore the widneing.
+            auto is_x_wide_int = op->type.is_int() && is_int(x, bits * 2);
+            auto is_x_wide_uint = op->type.is_uint() && is_uint(x, bits * 2);
+            auto is_x_wide_int_or_uint = is_x_wide_int || is_x_wide_uint;
+            // clang-format off
+            if (rewrite(max(min(rounding_shift_right(x, y), upper), lower), rounding_shift_right(x, y), is_x_wide_int_or_uint) ||
+                rewrite(rounding_shift_right(x, y), rounding_shift_right(x, y), is_x_wide_int_or_uint) ||
+                rewrite(rounding_shift_left(x, y), rounding_shift_left(x, y), is_x_wide_int_or_uint) ||
+                false) {
+                const Call *shift = Call::as_intrinsic(rewrite.result, {Call::rounding_shift_right, Call::rounding_shift_left});
```

Comment:
Is this all a janky workaround to not having something in IRMatch that just matches and gives you the wildcards instead of doing a full rewrite?

Reply:
I'd say this is a janky workaround for not being able to express lossless-cast like things in rewrite rules. What we're trying to do here is the following rewrites:

    rewrite(saturating_narrow(rounding_shift_right(widen(x), widen(y))),rounding_shift_right(x, y), can_prove(y > 0))
    rewrite(narrow(rounding_shift_right(widen(x), widen(y))),rounding_shift_right(x, y))
    rewrite(narrow(rounding_shift_right(widen(x), widen(y))),rounding_shift_right(x, y))

Doing this requires a combination of pattern matching and logic, which is in the body of the if.

Reply:
Added comment.